src/values/ObjectValue.js

```diff
@@ -645,25 +645,46 @@ export default class ObjectValue extends ConcreteValue {
     }
 
     if (!(P instanceof AbstractValue)) return this.$Get(P, Receiver);
+
+    // A string coercion might have side-effects.
+    // We assume that simple objects mean that they don't have a side-effectful
+    // valueOf and toString. (Is this enforced?)
+    if (P.mightNotBeString() && P.mightNotBeNumber() && !P.isSimpleObject()) {
+      if (this.$Realm.isInPureScope()) {
+        // If we're in pure scope, we can havoc the key and keep going.
+        // Coercion can only have effects on anything reachable from the key.
+        Havoc.value(this.$Realm, P);
+      } else {
+        AbstractValue.reportIntrospectionError(P, "property key might not have a well behaved toString");
```

Comment:
We do have diagnostic error codes for this sort of thing. We really really really should not introduce any new uses of reportIntrospectionError.

Reply:
Yea I'm pretty confused about best practices here. I wonder if we should do a pass over everything. It's hard to change a convention when the old one is still there.

Can you point to an example that is doing this correctly that I can follow?

Reply:
https://fburl.com/aqqtguux

Reply:
The main problem with a big code mod is that the existing uses of reportIntrospectionError are in places that are convenient for us, but which do not have the right source context at hand and do not facilitate recovery actions.